### PR TITLE
Add DMA functionality

### PR DESCRIPTION
**Pull request by RJ-Russell**
Implement DMA driver.

Tested by adding a task that uses the DMA to transmit data to the Usart and out to minicom.

@dnseitz

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -3,7 +3,7 @@ target = thumbv6m-none-eabi
 cargo-native = cargo
 cargo-target = xargo
 
-native-args = --features "test"
+native-args = --features "test serial"
```
Comment (dnseitz) [inline Makefile]: Any reason this was added?

File: src/exceptions/dma.rs [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+* This program is distributed in the hope that it will be useful,
+* but WITHOUT ANY WARRANTY; without even the implied warranty of
+* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+* GNU General Public License for more details.
+*
+* You should have received a copy of the GNU General Public License
+* along with this program. If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#![cfg(feature="dma")]
```
Comment (dnseitz) [inline src/exceptions/dma.rs]: This should be moved to the module definition in `src/exceptions/mod.rs` to stay consistent with how it was done for `serial`

File: src/io/mod.rs [hunk 3]
```diff
@@ -74,6 +77,33 @@ mod imp {
         ($fmt:expr, $($arg:tt)*) => (print!(concat!($fmt, "\n"), $($arg)*));
     }
 
+    struct DMASerial {
```
Comment (dnseitz) [inline src/io/mod.rs]: Moving all these implementations into it's own module seems like it would make the file a bit cleaner.

File: src/lib.rs [hunk 4]
```diff
@@ -117,8 +117,8 @@ pub fn init() -> ! {
     init_clock();
     init_ticks();
     init_usart();
-
-    unsafe { application_entry() };
+    init_dma();
+unsafe { application_entry() };
```
Comment (dnseitz) [inline src/lib.rs]: Reindent this line

File: src/peripheral/dma/ccr.rs [hunk 5]
```diff
@@ -0,0 +1,397 @@
...
+#[derive(Copy, Clone, Debug)]
+pub enum PeriphAndMemSize {
+    /// Eight bits.
+    Eight,
+    /// Sixteen bits.
+    Sixteen,
+    /// Thirty-two bits.
+    ThirtyTwo,
+    /// Reserved.
+    Reserved,
```
Comment (dnseitz) [inline src/peripheral/dma/ccr.rs]: Any reason to have a reserved variant? Seems like it's just reserved by hardware and you shouldn't be able to set it to that value.

**Review by dnseitz — CHANGES_REQUESTED** [review]

File: src/peripheral/dma/ccr.rs (hunk 5 above)
Comment (RJ-Russell) [reply src/peripheral/dma/ccr.rs]: Yes.
```Bits 11:10 MSIZE[1:0] Memory size
Bits 11:10 MSIZE[1:0]: Memory size
These bits are set and cleared by software.
    00: 8-bits
    01: 16-bits
    10: 32-bits
    11: Reserved 
```
There is an option for the setting so it should be included. As the description states, these bits are `set and cleared by software`.

Comment (dnseitz) [reply src/peripheral/dma/ccr.rs]: But Reserved generally means reserved by hardware, and that it's not a valid configuration. Correct me if it says explicitly that setting the bits to that value actually has some valid behavior.

Comment (RJ-Russell) [reply src/peripheral/dma/ccr.rs]: Yeah. Data sheet is kind of confusing. I removed the Reserved option. Let me know if there is anything else I should look at. Thanks.

**Review by dnseitz — APPROVED** [review]

**Pull request merged**